light-client-verifier/src/verifier.rs

```diff
@@ -111,119 +111,163 @@ where
+            self.predicates
+                .is_monotonic_height(untrusted.signed_header.header.height, trusted.height)?;
+        }
+
+        Ok(())
+    }
 
+    /// Check there is enough overlap between the validator sets of the trusted and untrusted
+    /// blocks.
+    pub fn verify_commit_trusting(
+        &self,
+        untrusted: &UntrustedBlockState<'_>,
+        trusted: &TrustedBlockState<'_>,
+        options: &Options,
+    ) -> Result<(), VerificationError> {
+        let trusted_next_height = trusted.height.increment();
+
+        if untrusted.height() != trusted_next_height {
             // Check there is enough overlap between the validator sets of
             // the trusted and untrusted blocks.
-            verdict!(self.predicates.has_sufficient_validators_overlap(
+            self.predicates.has_sufficient_validators_overlap(
                 untrusted.signed_header,
                 trusted.next_validators,
                 &options.trust_threshold,
                 &self.voting_power_calculator,
-            ));
+            )?;
         }
 
-        // Verify that more than 2/3 of the validators correctly committed the block.
-        verdict!(self.predicates.has_sufficient_signers_overlap(
-            untrusted.signed_header,
-            untrusted.validators,
-            &self.voting_power_calculator,
-        ));
+        Ok(())
+    }
+}
 
+impl<P, C, V, H> Verifier for PredicateVerifier<P, C, V, H>
+where
+    P: VerificationPredicates,
+    C: VotingPowerCalculator,
+    V: CommitValidator,
+    H: Hasher,
+{
+    /// Validate the given light block state.
+    ///
+    /// - Ensure the latest trusted header hasn't expired
+    /// - Ensure the header validator hashes match the given validators
+    /// - Ensure the header next validator hashes match the given next validators
+    /// - Additional implementation specific validation via `commit_validator`
+    /// - Check that the untrusted block is more recent than the trusted state
+    /// - If the untrusted block is the very next block after the trusted block, check that their
+    ///   (next) validator sets hashes match.
+    /// - Otherwise, ensure that the untrusted block has a greater height than the trusted block.
+    ///
+    /// **NOTE**: If the untrusted state's `next_validators` field is `None`,
+    /// this will not (and will not be able to) check whether the untrusted
+    /// state's `next_validators_hash` field is valid.
+    fn verify(
+        &self,
+        untrusted: UntrustedBlockState<'_>,
+        trusted: TrustedBlockState<'_>,
+        options: &Options,
+        now: Time,
+    ) -> Verdict {
+        verdict!(self.validate_untrusted(&untrusted));
+        verdict!(self.validate_trusting(&untrusted, &trusted, options, now));
+        verdict!(self.verify_commit(&untrusted));
+        verdict!(self.verify_commit_trusting(&untrusted, &trusted, options));
```

Comment:
To keep the original order that we had in `verify()`:
```suggestion
        verdict!(self.verify_commit_trusting(&untrusted, &trusted, options));
        verdict!(self.verify_commit(&untrusted));
```